### PR TITLE
Annotation: Optionally allow storing longer annotation tags

### PR DESCRIPTION
**Pull request by papagian**
**What this PR does / why we need it**:
Alters the `tags` column of the `annotation` table to `VARCHAR(4096)` (from  `VARCHAR(500)` before).
It introduces also a configuration `annotations.tags_length` for enforcing a lower limit (than the one imposed by the database) during saving/updating annotations. Its default value is `500`. However, changing it a much higher value is not recommended, because it can impact the performance.

**Which issue(s) this PR fixes**:

Fixes #51815

**Special notes for your reviewer**:
The respective [model field](https://github.com/grafana/grafana/blob/7779ae315b6e5e1217ffd38cc31a1f704bdd8440/pkg/services/annotations/annotations.go#L39) is `[]string` and it is stored in the respective `tags` column as a stringified array.
In addition to this column there is a `annotation_tag` table that keeps also each annotation tags. Hence, each tag is subject [of other database limits](https://github.com/grafana/grafana/blob/7779ae315b6e5e1217ffd38cc31a1f704bdd8440/pkg/services/sqlstore/migrations/tag_mig.go#L10).
I'm not completely sure whether the `annotation.tags` column is still used or if it's completely replaced by the `annotation_tag` table.

### CODE REVIEW

File: conf/defaults.ini
Comment (sakjur) [inline conf/defaults.ini:934]: Isn't 0 also possible?

File: pkg/services/annotations/annotations.go
Comment (sakjur) [inline pkg/services/annotations/annotations.go:15]: Make this more specific, it's better to have `ErrBaseTagLimitExceeded`, as you'd expect one unique message layout per base (for internationalization purposes). I guess in this case you probably also want to tell the caller that that was the cause, so add `errutil.WithPublicMessage` to it as well to make sure the user gets a good message.

You also want to switch out the `Error` in `pkg/api` for the caller to `ErrWithFallback` (literal word substitution there, they're compatible with each other)

Maybe something like this?

```suggestion
	ErrBaseTagLimitExceeded  = errutil.NewBase(errutil.StatusBadRequest, "annotations.tag-limit-exceeded", errutil.WithPublicMessage("Tags length exceeds the maximum allowed."))
```

File: pkg/services/sqlstore/migrations/annotation_mig.go
Comment (sakjur) [inline pkg/services/sqlstore/migrations/annotation_mig.go:184]: Ze juicy part.

This looks ok to me, but I'd appreciate a second pair of eyes on this or would like to set aside some time to benchmark it/browse MySQL/Postgres docs to figure out the risks of running this if you say have a few million annotations before I'm comfortable with approving this 🙈

In case anyone wonders, the reason why SQLite doesn't need a migration is because it has dynamic typing by default which we're using, so it doesn't actually care about limits on VARCHARs, see [datatypes in SQLite3](https://www.sqlite.org/datatype3.html)

> Note that numeric arguments in parentheses that following the type name (ex: "VARCHAR(255)") are ignored by SQLite - SQLite does not impose any length restrictions (other than the large global [SQLITE_MAX_LENGTH](https://www.sqlite.org/limits.html#max_length) limit) on the length of strings, BLOBs or numeric values.

**Comment by ryantxu** [comment]
Looking at the discussion now... why do we want annotations to be longer than 500 characters?

If that is a real use case, can they just write annotations to loki (or any other datasource that supports annotations)?

File: conf/defaults.ini
Comment (sakjur) [inline conf/defaults.ini:936]: QQ: Is there a performance problem with more than 500? Could we skip this conf?

**Comment by sakjur** [comment]
@ryantxu We received this as an ask via paid support and realized that the current limit was arbitrarily set to a fairly low limit and agreed to increasing it.

Do you think it would be a problem for us to do this? Could you expand a little bit on why if so?

File: conf/defaults.ini
Comment (papagian) [reply conf/defaults.ini:934]: right now it is but maybe it does not make much sense; I think that we have to allow only values between (500, 4096). WDYT?

File: pkg/services/annotations/annotations.go
Comment (papagian) [reply pkg/services/annotations/annotations.go:15]: > Make this more specific, it's better to have ErrBaseTagLimitExceeded, as you'd expect one unique message layout per base (for internationalization purposes).
GTK, I had the impression that it should be one base per status code.

> I guess in this case you probably also want to tell the caller that that was the cause, so add errutil.WithPublicMessage to it as well to make sure the user gets a good message.
If we provide a public message do we still need [the custom](https://github.com/grafana/grafana/blob/ceed68e4c4f749f14cdd8d5880b08022a9a95689/pkg/services/sqlstore/annotation.go#L393) one?

File: pkg/services/sqlstore/migrations/annotation_mig.go
Comment (papagian) [reply pkg/services/sqlstore/migrations/annotation_mig.go:184]: the SQL considerations are also discussed [here](https://github.com/grafana/grafana/discussions/51815)

I'll trying to monitor its performance but we have run [similar migrations](https://github.com/grafana/grafana/blob/17fc5d82d32c080dceeeec02837f7f09e2944d98/pkg/services/sqlstore/migrations/dashboard_mig.go#L94) for other tables

File: pkg/services/annotations/annotations.go
Comment (sakjur) [reply pkg/services/annotations/annotations.go:15]: > If we provide a public message do we still need [the custom](https://github.com/grafana/grafana/blob/ceed68e4c4f749f14cdd8d5880b08022a9a95689/pkg/services/sqlstore/annotation.go#L393) one?

Yes, that message is for Go, this message is for end-users (it might seem like it's the same thing, but we essentially want to start separating those because the contents of what'd be a good error message is different depending on if you're facing the end-user or an operator/developer)

File: pkg/services/sqlstore/migrations/annotation_mig.go
Comment (papagian) [reply pkg/services/sqlstore/migrations/annotation_mig.go:184]: I have created 500k annotations using generators following [this](https://github.com/grafana/grafana/pull/21915#issuecomment-582328787) and I was able to run the migration without issues.

File: conf/defaults.ini
Comment (papagian) [reply conf/defaults.ini:950]: I don't know; I got the impression that you had suggested it 😳

**Review by ying-jeanne — APPROVED** [review]
LGTM, I looked through the comments and let you decide whether you want another around review from Emil just to be 💯 sure.

File: pkg/services/annotations/annotationsimpl/xorm_store.go
Comment (sakjur) [inline pkg/services/annotations/annotationsimpl/xorm_store.go:395]: This is definitely something that shouldn't affect merging this, as the only consequence of getting this wrong would be certain border values getting a less helpful error message.

I think [Xorm's type conversion](https://gitea.com/xorm/xorm/src/tag/v0.8.2/session_convert.go#L643-L649) uses json.Marshal here, so it's possible that this validation will pass but the later validation fail because of the difference between Go's `[a b c]` and JSON's `["a", "b", "c"]` 🤔

Presumably, another way of getting an estimate of the length without having to convert it to JSON would be something like:

```
estimatedTagLength := len(item.Tags) * 4 // each item has approximately a pair of quotes, a comma and a space for each object except for the last, plus two brackets.
for _, t := range item.Tags {
	estimatedTagLength += len(t)
}
```

File: pkg/setting/setting.go
Comment (sakjur) [inline pkg/setting/setting.go:614]: I believe we can tone this message down a little bit to avoid scaring the user.

```suggestion
		cfg.Logger.Info("[annotations.tags_length] has been increased from its default value; this may affect the performance", "tagLength", cfg.AnnotationMaximumTagsLength)
```

**Review by sakjur — APPROVED** [review]
LGTM

File: pkg/services/annotations/annotationsimpl/xorm_store.go
Comment (papagian) [reply pkg/services/annotations/annotationsimpl/xorm_store.go:395]: good point, I will fix it

**Pull request merged**
